dbms/src/Storages/Transaction/FastAddPeer.cpp

```diff
@@ -50,382 +51,267 @@ FastAddPeerContext::FastAddPeerContext(uint64_t thread_count)
     tasks_trace = std::make_shared<AsyncTasks>(thread_count);
 }
 
-TempUniversalPageStoragePtr FastAddPeerContext::getTempUniversalPageStorage(UInt64 store_id, UInt64 upload_seq)
+ParsedCheckpointDataHolderPtr FastAddPeerContext::CheckpointCacheElement::getParsedCheckpointData()
 {
-    std::unique_lock lock(ps_cache_mu);
-    auto iter = temp_ps_cache.find(store_id);
-    if (iter != temp_ps_cache.end() && iter->second.first >= upload_seq)
-    {
-        return iter->second.second;
-    }
-    return nullptr;
+    std::unique_lock lock(mu);
+    if (!is_ready)
+        cv.wait(lock, [&] { return is_ready; });
```

Comment:
I don't think we need a `condition_variable` here. It is complicated and dangerous. For example, `setParsedCheckpointData` could throw exceptions, then `is_ready` won't never be true.

A much simpler way is using a lock to protect a method named `parseCheckpointData`:

```
lock(mutext);
if parsed_checkpoint_data exist
then 
    return parsed_checkpoint_data;
else
   do parse job;
```

Reply:
Fixed.